### DISCUSSION TITLE
How to get the extended metrics per each transaction/test step in aggregate way?

### DISCUSSION THREAD

**Discussion by rhiadechavez** [artilleryio/artillery#2250] [Artillery]
Hello,

I am very enthusiastic about using artillery-engine-playwright. 
Currently, I am conducting a proof of concept for our team and I am deeply committed to understand and mastering its usage.

I have created a sample script below and have been experimenting with a light load on this test environment. 
I was successful in obtaining performance metrics in an aggregate manner for each transaction, as reflected in the sample summary report below.

However, I am now curious if it's possible to obtain extended metrics such as CLS, FCP, LCP, TTFB, domcontentloaded, dominteractive, and memory_used_mb for each transaction. Is there a method to achieve this?

Thank you for your assistance.

Here's sample js script.

```javascript
module.exports = { PetStore };
const { expect } = require("@playwright/test");


const withTransactionTimer = async (transactionName, events, userActions) => {
    const startedTime = Date.now();
    await userActions();
    const difference = Date.now() - startedTime;
  
    events.emit('histogram', transactionName, difference)
  }


async function PetStore(page, userContext, events) {

    await withTransactionTimer("SC01_PetStore_Retail_SignUp_T01_Launch", events, async () => {
        await page.goto('https://petstore.octoperf.com/actions/Catalog.action');
    })

    const content = await page.content();
    //console.log(content);
    let jsessionID = content.match(';jsessionid=(.*?)">');
    
//console.log(jsessionID[1]);
  
    await withTransactionTimer("SC02_PetStore_Retail_T02_Login", events, async () => {
        await page.locator('text=Sign In').click();
        await page.waitForURL(`https://petstore.octoperf.com/actions/Account.action;jsessionid=${jsessionID[1]}?signonForm=`);
        await page.locator('input[name="username"]').click();
        await page.locator('input[name="username"]').fill('juandelacruz');
        await page.locator('input[name="password"]').click();
        await page.locator('input[name="password"]').fill('12345');
        await page.locator('input[name="password"]').press('Enter');

        await page.waitForURL('https://petstore.octoperf.com/actions/Catalog.action');
    //await expect(locator).toContainText('text=Welcome Juan!'); //Connection details
        await expect(page.getByText('Welcome Juan!')).toBeVisible();

    })
  
    
    await withTransactionTimer("SC03_PetStore_Retail_T03_Checking_Items", events, async () => { 
      
        await page.locator('#SidebarContent').getByRole('link').first().click();
        await page.getByRole('link', { name: 'Return to Main Menu' }).click();
  
        for (let i = 1; i < 5; i++){
  
            await page.locator('#SidebarContent').getByRole('link').nth(`${i}`).click();
            await page.getByRole('link', { name: 'Return to Main Menu' }).click();
  
      }
    })
  
   
    await withTransactionTimer("SC04_PetStore_Retail_T04_Get_Birds", events, async () => { 
  
        await page.locator('#SidebarContent').getByRole('link').nth(4).click();

        await expect(page.getByText('Amazon Parrot')).toBeVisible();
        //await page.getByRole('cell', { name: 'Amazon Parrot' }).click();
        await page.getByRole('link', { name: 'AV-CB-01' }).click();
        await page.getByRole('link', { name: 'Add to Cart' }).click();
        await page.getByRole('link', { name: 'Return to Main Menu' }).click();
    
     })
  
    
    await withTransactionTimer("SC05_PetStore_Retail_T05_Get_Fish", events, async () => {     
  
        await page.locator('#SidebarContent').getByRole('link').first().click();
        await expect(page.getByText('Angelfish')).toBeVisible();
      //await page.getByRole('cell', { name: 'Amazon Parrot' }).click();
      //await expect(page).toHaveTitle("Fish");
        await page.getByRole('link', { name: 'FI-SW-01' }).click();
        await page.getByRole('link', { name: 'Add to Cart' }).first().click();
        await page.getByRole('link', { name: 'Return to Main Menu' }).click();
    
    })
    
    
    
    await withTransactionTimer("SC06_PetStore_Retail_T06_Get_Dog_From_Search", events, async () => {     
  
        await page.locator('input[name="keyword"]').click();
        await page.locator('input[name="keyword"]').fill('Golden');
        await page.locator('text=Search').click();
        await expect(page.getByText('Golden Retrieve')).toBeVisible();
        await expect(page.getByText('Great family dog')).toBeVisible();
        await page.locator('td:has-text("K9-RT-01")').click();
        
        await page.waitForTimeout(1000); 
        
        await page.locator('text=Add to Cart').click();
        //await page.getByRole('link', { name: 'Add to Cart' }).click();
        await page.getByRole('link', { name: 'Return to Main Menu' }).click();
    
    })
    

    
    await withTransactionTimer("SC07_PetStore_Retail_T07_Proceed_To_Checkout", events, async () => {     
  
        await page.locator('#MenuContent').getByRole('link').first().click();
        await page.getByRole('link', { name: 'Proceed to Checkout' }).click();
        await page.waitForURL('https://petstore.octoperf.com/actions/Order.action?newOrderForm=');
    })
  
  
    
    await withTransactionTimer("SC08_PetStore_Retail_T08_Fill_Out_CheckOutForm", events, async () => {     
  
        await page.locator('select[name="order\\.cardType"]').selectOption('MasterCard');
        await page.locator('input[name="order\\.creditCard"]').fill('999 9999 9999 9999');
        await page.locator('input[name="order\\.billToFirstName"]').fill('Juan');
        await page.locator('input[name="order\\.expiryDate"]').fill('12/03');
        await page.locator('input[name="order\\.billToLastName"]').fill('Dela Cruz');
        await page.locator('input[name="order\\.billAddress1"]').fill('Street St');
        await page.locator('input[name="order\\.billAddress2"]').fill('Street St');
        await page.locator('input[name="order\\.billCity"]').fill('Auckland');
        await page.locator('input[name="order\\.billState"]').fill('Auckland');
        await page.locator('input[name="order\\.billZip"]').fill('1121');
        await page.locator('input[name="order\\.billCountry"]').fill('New Zealand');
        await page.locator('text=Continue').click();
        await page.waitForURL('https://petstore.octoperf.com/actions/Order.action');
  
    })
  
  
    const content2 = await page.content();
    //console.log(content2);
    let dateTime = content2.match('<b> (.*?)</b></font>');
    //console.log(dateTime[1]);
    
    
    
    await withTransactionTimer("SC09_PetStore_Retail_T09_Confirm_CheckOut", events, async () => {     
  
        await expect(page.getByText(`Please confirm the information below and then press continue... Order ${dateTime[1]}`)).toBeVisible();
    
        await page.locator('text=Confirm').click();
        await page.waitForURL('https://petstore.octoperf.com/actions/Order.action?newOrder=&confirmed=true');
        await expect(page.getByText('Thank you, your order has been submitted.')).toBeVisible();
        await page.getByRole('link', { name: 'Return to Main Menu' }).click();
    })
  
  
  };
  //);
``` 

Here's summary report from artillery. 

Summary report @ 19:19:17(+1300)
--------------------------------

SC01_PetStore_Retail_SignUp_T01_Launch:
  min: ......................................................................... 2329
  max: ......................................................................... 2329
  median: ...................................................................... 2322.1
  p95: ......................................................................... 2322.1
  p99: ......................................................................... 2322.1
SC02_PetStore_Retail_T02_Login:
  min: ......................................................................... 1888
  max: ......................................................................... 1888
  median: ...................................................................... 1901.1
  p95: ......................................................................... 1901.1
  p99: ......................................................................... 1901.1
SC03_PetStore_Retail_T03_Checking_Items:
  min: ......................................................................... 5184
  max: ......................................................................... 5184
  median: ...................................................................... 5168
  p95: ......................................................................... 5168
  p99: ......................................................................... 5168
SC04_PetStore_Retail_T04_Get_Birds:
  min: ......................................................................... 2149
  max: ......................................................................... 2149
  median: ...................................................................... 2143.5
  p95: ......................................................................... 2143.5
  p99: ......................................................................... 2143.5
SC05_PetStore_Retail_T05_Get_Fish:
  min: ......................................................................... 2053
  max: ......................................................................... 2053
  median: ...................................................................... 2059.5
  p95: ......................................................................... 2059.5
  p99: ......................................................................... 2059.5
SC06_PetStore_Retail_T06_Get_Dog_From_Search:
  min: ......................................................................... 3165
  max: ......................................................................... 3165
  median: ...................................................................... 3134.5
  p95: ......................................................................... 3134.5
  p99: ......................................................................... 3134.5
SC07_PetStore_Retail_T07_Proceed_To_Checkout:
  min: ......................................................................... 1051
  max: ......................................................................... 1051
  median: ...................................................................... 1043.3
  p95: ......................................................................... 1043.3
  p99: ......................................................................... 1043.3
SC08_PetStore_Retail_T08_Fill_Out_CheckOutForm:
  min: ......................................................................... 798
  max: ......................................................................... 798
  median: ...................................................................... 804.5
  p95: ......................................................................... 804.5
  p99: ......................................................................... 804.5
SC09_PetStore_Retail_T09_Confirm_CheckOut:
  min: ......................................................................... 1019
  max: ......................................................................... 1019
  median: ...................................................................... 1022.7
  p95: ......................................................................... 1022.7
  p99: ......................................................................... 1022.7
browser.http_requests: ......................................................... 368
browser.memory_used_mb:
  min: ......................................................................... 1.6
  max: ......................................................................... 17.2
  median: ...................................................................... 8.9
  p95: ......................................................................... 15.3
  p99: ......................................................................... 16
browser.page.CLS.Pet Store Retail:
  min: ......................................................................... 0
  max: ......................................................................... 0.1
  median: ...................................................................... 0
  p95: ......................................................................... 0
  p99: ......................................................................... 0
browser.page.FCP.Pet Store Retail:
  min: ......................................................................... 431.7
  max: ......................................................................... 2012.7
  median: ...................................................................... 450.4
  p95: ......................................................................... 925.4
  p99: ......................................................................... 2018.7
browser.page.LCP.Pet Store Retail:
  min: ......................................................................... 431.7
  max: ......................................................................... 2329.4
  median: ...................................................................... 450.4
  p95: ......................................................................... 925.4
  p99: ......................................................................... 925.4
browser.page.TTFB.Pet Store Retail:
  min: ......................................................................... 385.1
  max: ......................................................................... 1481.7
  median: ...................................................................... 407.5
  p95: ......................................................................... 871.5
  p99: ......................................................................... 871.5
browser.page.domcontentloaded: ................................................. 29
browser.page.domcontentloaded.Pet Store Retail: ................................ 29
browser.page.dominteractive:
  min: ......................................................................... 411
  max: ......................................................................... 1526
  median: ...................................................................... 441.5
  p95: ......................................................................... 528.6
  p99: ......................................................................... 907
browser.page.dominteractive.Pet Store Retail:
  min: ......................................................................... 411
  max: ......................................................................... 1526
  median: ...................................................................... 441.5
  p95: ......................................................................... 528.6
  p99: ......................................................................... 907
vusers.completed: .............................................................. 1
vusers.created: ................................................................ 1
vusers.created_by_name.Pet Store Retail: ....................................... 1
vusers.failed: ................................................................. 0
vusers.session_length:
  min: ......................................................................... 26682.4
  max: ......................................................................... 26682.4
  median: ...................................................................... 26643.2
  p95: ......................................................................... 26643.2
  p99: ......................................................................... 26643.2

**Comment by bernardobridge** [comment]
Hi @rhiadechavez 👋,

I'm glad to hear you're finding `artillery-engine-playwright` useful! 

I also see you're using the `withTransactionTimer` helper that I gave as a suggestion to another user. We actually now make this available for you by default since the new version, [Artillery v2.0.0-38](https://github.com/artilleryio/artillery/releases/tag/v2.0.0-38). You can read more about the `step()` helper here: https://www.artillery.io/docs/reference/engines/playwright#teststep-argument.

---

To answer your actual question, it's not currently possible to have an aggregate of those metrics per step, because the two concepts have different goals: 

## Web Vital Metrics / Extended Metrics
Web vital metrics like `TTFB`, `FCP` (or other extended metrics like `domcontentloaded`) etc, are conceptually tied to a page (which is why we list them as `browser.page`). They try to measure how fast a page feels to a real user. You can almost think of it as a [Google Lighthouse](https://developer.chrome.com/docs/lighthouse/overview/) report for each page your test visited (but at scale) - and in a Lighthouse score, you'd also be doing it `per page` by providing a URL. It's just that Artillery figures out the pages visited for you, and does these metrics (web vitals) at scale, for every virtual user.

## Steps (Transaction Timers)
These are intended to capture transactions which may take place within a page, and even after the page loads (this is why in the new convenience helper I mentioned in the beginning, we emit these metrics with the prefix `browser.step`, to differentiate from `page` metrics). For example, in your test, I believe in the step `SC06_PetStore_Retail_T06_Get_Dog_From_Search`, you fill in a search bar to make a search. You may be interested in finding out how long that takes, because, for instance, that makes a call to your backend. But that doesn't necessarily navigate to a new page.

---

It does seem like you're aggregating the web vital and extended metrics by scenario (`aggregateByName: true` setting). You might not want to do that, if you're interested in getting all the individual metrics **per page visited**. This would then allow you to use the two concepts I mentioned above together: 
- Web vitals would give you how the page "feels" when loading to the user, per page;
- Steps would give you additional timing information for any transaction you care about that might be happening within pages of your test.

Hopefully that helps, and if that doesn't make sense, please let us know! 🙇‍♂️  We're always interested in finding out other use-cases to inform how we're building the tool.

**Reply by rhiadechavez** [reply]
Hi @bernardobridge,

I wanted to express my gratitude for your prompt response – I truly appreciate it. 
Your explanation has been helpful in clarifying how the Web Vital metrics and the transaction timer should be utilized and how they function within Artillery.

And I completely agree with you; I should indeed set `aggregateByName: False` so that I can obtain all individual metrics. 
It works like a charm! 👍 💯 

Additionally, I have a quick question: does Artillery support running multiple scenarios in a single YAML file?

like something like this below. 

```dockerfile
scenarios:
- name: "Name 1 "
  engine: playwright
  flowFunction: "Launch"

- name: "Name 2"
  engine: playwright
  flowFunction: "Login"

- name: "Name 3"
  engine: playwright
  flowFunction: "CheckingItems
  
```

**Reply by bernardobridge** [reply]
Hey @rhiadechavez ! No problem at all, I'm glad it helped!

Artillery does support running multiple scenarios in a single YAML file. However, one thing to note is that they will not **all run**, but instead each virtual user will choose a scenario to run at random, based on weights (read more about it here: https://www.artillery.io/docs/reference/test-script#scenario-weights). If no `weight` is specified (as in your example), it defaults to 1. This is primarily used to represent virtual users coming into your application and doing different things in a weighted way (e.g. 30% of users do this, 50% do another thing, 20% do something else).

Since [Artillery v2.0.0-38](https://github.com/artilleryio/artillery/releases/tag/v2.0.0-38), we also support passing in the `--scenario-name` flag, which allows you to specify that only one of those scenarios should be run (by its name).

Hope that helps!

---

_P.S. `aggregateByName` defaults to false, so no need to set it explicitly if it's false._
